python-spec/src/somacore/data.py

```diff
@@ -65,7 +76,48 @@ def read(
     ) -> "ReadIter[pa.Table]":
         """Reads a user-defined slice of data into Arrow tables.
 
-        TODO: Further per-param documentation.
+        :param coords: for each index dimension, which rows to read.
+            Defaults to ``None``, meaning no constraint -- all IDs.
+        :param column_names: the named columns to read and return.
+            Defaults to ``None``, meaning no constraint -- all column names.
+        :param partitions: an optional ``ReadPartitions`` hint to indicate
+            how results should be organized.
+        :param result_order: order of read results.
+            This can be one of 'row-major', 'col-major', or 'auto'.
+        :param value_filter: an optional [value filter] to apply to the results.
+            Defaults to no filter.
+
+        **Indexing:**
+
+        Indexing is performed on a per-column basis for each indexed column.
+        To specify dimensions:
+
+        - A sequence of coordinates is accepted, one per indexed dimension.
+        - The sequence length must be less than or equal to the number of
+          indexed dimensions.
+        - If the sequence is shorter than the number of indexed coordinates,
+          then no constraint (i.e. ``None``) is used for the remaining
+          indexed dimensions.
+        - Specifying ``None`` or an empty sequence (e.g. ``()``) represents
+          no constraints over any dimension, returning the entire dataset.
+          TODO: https://github.com/single-cell-data/TileDB-SOMA/pull/910
+
+        Each dimension may be indexed as follows:
+
+        - ``None`` or ``slice(None)`` places no constraint on the dimension.
+        - Coordinates can be specified as a scalar value, a Python sequence
+          (``list``, ``tuple``, etc.), a ``ndarray``, an Arrow array, and
+          similar objects (as defined by ``SparseDFCoords``).
+        - Slices are doubly inclusive: ``slice(2, 4)`` means ``[2, 3, 4]``,
+          not ``[2, 3]``.  Slice *steps* may not be used: ``slice(10, 20, 2)``
+          is invalid.  ``slice(None)`` places no constraint on the dimension.
+          Half-specified slices like ``slice(None, 99)`` and ``slice(5, None)``
+          specify all indices up to and including that value, and all indices
+          starting from and including the value.
+        - Negative indexing is not supported.
+          TODO: What if the domain includes negative numbers?
```

Comment:
we forbid that

https://github.com/single-cell-data/TileDB-SOMA/issues/519

and if that isn't explicit here in the spec it should be

also -- discussion at the time with @atolopko-czi @aaronwolen @bkmartinjr and perhaps others:

* initially we wanted `soma_joinid` to be `uint64` so negative would be impossible
* that works in Python but for R we couldn't do unsigned so we went with `int64` with forbidding the negative side

I'm happy to take this as an action item on a separate PR to be clearer about forbidden negative indices in this spec repo -- just lmk if you don't want to do that on this PR (which is fine with me) & i'll open a separate PR

Reply:
Actually I went ahead and created #121 -- if you do that work on this PR, fanastic & you can close the issue -- otherwise we can do a follow-on PR.

Reply:
Given the [discussion on bug 121](https://github.com/single-cell-data/SOMA/issues/121#issuecomment-1426485914), I think this probably warrants reinstating language around negative indexing:

> Negative numbers as indices or parts of slices are used as raw domain values and are not treated specially, unlike traditional Python sequence indexing. For instance, `slice(-10, 3)` indicates the range from −10 to 3 on the given dimension.

This particular bit of language only belongs in DataFrame, since in the NDArrays, negative indices do not appear at all. (And if somebody tries to index their DataFrame’s `soma_join_id` by `[-10:3]`, then that’s fine; they’ll get indices 0–3 inclusive as if they did so on any other non-negative dimension.)